### DISCUSSION TITLE
Java version conflicts in janusgraph-full distribution

### DISCUSSION THREAD

**Discussion by porunov** [JanusGraph/janusgraph#3467] [General]
@JanusGraph/committers @JanusGraph/maintainers I wanted to discuss Java version conflicts in janusgraph-full build.

I'm currently working on the PR to add ElasticSearch 8 support here: #3268 

I bumped into the problem with janusgraph-dist tests failing for tests using ElasticSearch. After a small investigation it appears that ElasticSearch Server 8 doesn't support Java 8 or Java 11 anymore. ElasticSearch Server 8.6.0 requires Java 17, 18, or 19 ([see compatibility matrix](https://www.elastic.co/support/matrix#matrix_jvm)).
_Notice, ElasticSearch Rest client which we are using supports Java 8. It's only the server itself requires Java 17 or higher._

Our janusgraph-dist tests are running against Java 8 and Java 11 and then produce Java 8 and Java 11 compatible builds.
Gremlin Console supports Java 8 and Java 11, but not Java 17. 
JanusGraph server supports Java 8, Java 11, Java 17 (and probably more).

As for now, I decided to lock ElasticSearch 7.17.8 version in `janusgraph-dist` builds to be able to continue produce Java 8 and Java 11 compatible builds. 
There are possible other solutions I see to this problem:
- Remove janusgraph-full distributions
- Change JanusGraph Server script a little bit by providing `ELASTICSEARCH_JAVA_HOME` environment variable which gets Java from `ELASTICSEARCH_JAVA_HOME` instead of `JAVA_HOME`. In such case Cassandra, JanusGraph Server, Gremlin Console will use `JAVA_HOME` (which is either 8 or 11) and ElasticSearch will use `ELASTICSEARCH_JAVA_HOME` (which is 17, 18, or 19). I guess users might be confused a little bit about `ELASTICSEARCH_JAVA_HOME` environment variable, but if we document it in the [Installation docs](https://docs.janusgraph.org/getting-started/installation/#using-the-pre-packaged-distribution) then it should be good I think.
- Run janusgraph-dist builds using Java 17 only, but modify Gremlin Console a bit to introduce Java environment variable to the Gremlin Console (like `GREMLIN_JAVA_HOME`or something like that). In such case Cassandra, ElasticSearch, JanusGraph Server will be running using Java 17, but users will be opening Gremlin Console using Java 8 or Java 11. Again, we will need to document it.

Maybe there are other solutions I missed, but would be good to hear your thoughts or preferences here.

**Comment by farodin91** [comment]
I'm pretty sure at the moment elasticsearch has their own Java distribution.
